Add new extension methods for multi dimensional array

hctrdev opened the pull request.

Commit:
Add a new Any extension method for multi-dimension arrays

Add a new extension method for multi-dimensional arrays to check if any
value in the array matches a given condition.
Add a test for the new extension method.

Commit:
Add the new method to the 2D array interface

Update the 2D array interface to also expose the new Any function. The
implementation of the new method on the 2D array delegates to the new
multi-dimensional array extension.

Commit:
Add new All extension

Add a new extension to check that all values in an array match a given
condition. The All functionality leverages the Any functionality to avoid
duplicating code.
Add test for the new extension and update existing test to use a data
source instead of hard-coding test cases in the test method itself.

Commit:
Split interfaces for one-based arrays into separate classes

Commit:
Add Count extension to multi-dim array

Add a new Count extension to the multi-dimensional array and 2D one based
array to allow counting the number of elements in an array that match the
provided function.
Add test to ensure the count works as expected.

Commit:
Add slicing methods to one based arrays

Update the interfaces for OneBasedArrays to include the methods to slide
a 2D array to a single row/column. These methods delegate down to the
existing methods for standard arrays.

Commit:
Fix 2D array slicing methods

Add test for the new code slicing 2D arrays into rows and columns to
ensure the sliced data has the correct length.
Fix bug caused by mixing one-based array indexing with zero-based.

Commit:
Add method to convert single dimension array to 2D array

Add a new method on the one-based array interface to convert a single
dimension array to a 2D array. This method uses the underlying extension
method on the raw array.
Add test for the new method.

Commit:
Add new methods for collapsing arrays

Add new extension methods for the single and multi-dimensional arrays to
collapse arrays down either to a single value or to a row/column.
Add methods on the one-based array interface and class that delegate down
to the array extensions.
Add new tests to ensure the new methods work as expected.

Commit:
Simplify code collapsing 2D arrays

Update the code collapsing 2D arrays to a single row/column to split the
array into rows/columns and then call the existing code that collapses a
1D array to a single value. This makes the code easier to understand.
Add documentation to new one-based array methods.

ColmBhandal commented on CsharpExtras/Enumerable/OneBased/OneBasedArray.cs:
```diff
@@ -172,5 +142,15 @@ public IOneBasedArray<TResult> Map<TResult>(Func<TVal, TResult> mapper)
+            return new OneBasedArray2DImpl<TVal>(ZeroBasedEquivalent.To2DArray(orientation));
+        }
+
+        public TVal CollapseToSingleValue(Func<TVal, TVal, TVal> collapseFunction)
```
Rename collapse to fold

ColmBhandal reviewed (changes requested):
See comment

Commit:
Rename collapse methods to fold

Update the naming of methods from Collapse to Fold to follow the standard
naming scheme for this operation.
